lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorState.java

```diff
@@ -22,17 +22,150 @@
+  FAILED,
+
+  /** Unclean shutdown. */
+  KILLED;
+
+  /**
+   * Check if evaluator is in the initial state (ALLOCATED).
+   * @return true if allocated, false otherwise.
+   */
+  public final boolean isAllocated() {
+    return this == ALLOCATED;
+  }
+
+  /**
+   * Check if evaluator is in SUBMITTED state.
+   * @return true if submitted, false otherwise.
+   */
+  public final boolean isSubmitted() {
+    return this == SUBMITTED;
+  }
+
+  /**
+   * Check if the evaluator is in running state.
+   * @return true if RUNNING, false otherwise.
+   */
+  public final boolean isRunning() {
+    return this == RUNNING;
+  }
+
+  /**
+   * Check if the evaluator is in the process of being shut down.
+   * @return true if evaluator is being closed, false otherwise.
+   */
+  public final boolean isClosing() {
+    return this == CLOSING;
+  }
+
+  /**
+   * Check if evaluator is in one of the active states (ALLOCATED, SUBMITTED, or RUNNING).
+   * @return true if evaluator is available, false if it is closed or in the process of being shut down.
+   */
+  public final boolean isAvailable() {
+    return this == ALLOCATED || this == SUBMITTED || this == RUNNING;
+  }
+
+  /**
+   * Check if the evaluator is stopped. That is, in one of the DONE, FAILED, or KILLED states.
+   * @return true if evaluator completed, false if it is still available or in the process of being shut down.
+   */
+  public final boolean isCompleted() {
+    return this == DONE || this == FAILED || this == KILLED;
+  }
+
+  /**
+   * Check if the evaluator is closed due to an error. That is, in FAILED or KILLED state.
+   * @return true if evaluator is stopped due to an error, true otherwise.
+   */
+  public final boolean isCompletedAbnormally() {
+    return this == FAILED || this == KILLED;
+  }
+
+  /**
+   * Check if transition from current state to the given one is legal.
+   * @param toState new state to transition to.
+   * @return true if transition is legal, false otherwise.
+   */
+  public final boolean isLegalTransition(final EvaluatorState toState) {
+
+    if (this == toState) {
+      return true;
+    }
+
+    switch(this) {
+
+    case ALLOCATED:
+      switch(toState) {
+      case SUBMITTED:
+      case CLOSING:
+      case DONE:
+      case FAILED:
+      case KILLED:
+        return true;
+      default:
+        return false;
+      }
+
+    case SUBMITTED:
+      switch(toState) {
+      case RUNNING:
+      case CLOSING:
+      case DONE:
+      case FAILED:
+      case KILLED:
+        return true;
+      default:
+        return false;
+      }
+
+    case RUNNING:
+      switch(toState) {
+      case CLOSING:
+      case DONE:
+      case FAILED:
+      case KILLED:
+        return true;
+      default:
+        return false;
+      }
+
+    case CLOSING:
+      switch(toState) {
+      case DONE:
+      case FAILED:
+      case KILLED:
+        return true;
+      default:
+        return false;
+      }
+
+    default:
+      return false;
```

Comment:
Behavior change, doesn't check for all considered states and throw by default now